fxtest/fxtest_test.go

```diff
@@ -154,4 +158,27 @@ func TestLifecycle(t *testing.T) {
 		lc.RequireStop()
 		assert.Equal(t, 1, spy.failures, "Expected lifecycle stop to fail.")
 	})
+
+	t.Run("StopLeaks", func(t *testing.T) {
+		spy := newTB()
+		lc := NewLifecycle(spy)
+		// Ensure we do not leak a context goroutine and hide application Stop errors
+		// since the return statement below would never be reached if we are leaking the context
+		funcRunForever := func(ctx context.Context) error {
+			<-ctx.Done()
+			return ctx.Err()
+		}
+
+		lc.Append(fx.Hook{
+			OnStart: func(ctx context.Context) error { go funcRunForever(ctx); return nil },
+			OnStop:  func(ctx context.Context) error { return nil },
+		})
+
+		lc.RequireStart()
+		assert.Equal(t, 0, spy.failures, "Expected lifecycle start to succeed.")
+
+		lc.RequireStop()
+		assert.Equal(t, 0, spy.failures, "Expected lifecycle to stop.")
+	})
```

Comment:
I'd like to verify that we're not leaking extra goroutines (beyond what the user intentionally starts) between `RequireStart` and `RequireStop`. To do that, we'll need to rework this this to look a little more like a production Fx app that starts a long-running workload in `OnStart`.

In this case, I think we'll want to have `OnStart` close over two signals channel and kick off a goroutine. In `OnStop`, we signal the `OnStart` goroutine to stop and block until it shuts down.

```
stop := make(chan struct{})
stopped := make(chan struct{})

onStart := func(ctx context.Context) error {
  go func() { 
    <-stop
    close(stopped)
  }()
  return ctx.Err()
}

onStop := func(ctx context.Context) error {
  close(stop)
  <-stopped
  return ctx.Err()
}

lc.Append(fx.Hook{
  OnStart: onStart,
  OnStop: onStop,
})
```

This should let us verify that only the anonymous function in `onStart` was leaked between app startup and shutdown.

Reply:
I updated the test, i was not sure if you intended the `ctx.Done` goroutine to be a part of the `onStart` function but this without the changes does in fact leak.